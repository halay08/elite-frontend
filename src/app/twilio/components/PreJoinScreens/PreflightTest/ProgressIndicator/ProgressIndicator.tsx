import React from 'react';
import { makeStyles } from '@material-ui/core';
import { TEST_DURATION } from '../PreflightTest';

const SIZE = 24;
const THICKNESS = 3;
const OUTER_RADIUS = SIZE / 2;
const INNER_RADIUS = OUTER_RADIUS - THICKNESS;
const INNER_CIRCUMFERENCE = 2 * Math.PI * INNER_RADIUS;

const useStyles = makeStyles(() => ({
  svg: {
    transform: 'rotate(-90deg)',
    marginRight: '0.6em',
    fill: 'none',
  },
  innerCircle: {
    stroke: 'rgb(200, 204, 207)',
  },
  circle: {
    stroke: '#027AC5',
    strokeDasharray: INNER_CIRCUMFERENCE,
    strokeDashoffset: '0px',
    animation: `${TEST_DURATION / 1000 + 1}s $progress linear`,
  },
  '@keyframes progress': {
    '0%': {
      strokeDashoffset: `${INNER_CIRCUMFERENCE}px`,
    },
    '100%': {
      strokeDashoffset: '0px',
    },
  },
}));

export default function ProgressIndicator() {
  const classes = useStyles();

  return (
    <svg
      viewBox={`0 0 ${SIZE} ${SIZE}`}
      width={SIZE}
      height={SIZE}
      strokeWidth={`${THICKNESS}px`}
      xmlns="http://www.w3.org/2000/svg"
      className={classes.svg}
    >
      <circle
        cx={OUTER_RADIUS}
        cy={OUTER_RADIUS}
        r={INNER_RADIUS}
        className={classes.innerCircle}
      />
      <circle
        cx={OUTER_RADIUS}
        cy={OUTER_RADIUS}
        r={INNER_RADIUS}
        className={classes.circle}
      />
    </svg>
  );
}
