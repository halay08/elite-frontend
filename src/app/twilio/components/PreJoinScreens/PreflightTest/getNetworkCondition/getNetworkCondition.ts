import { NetworkCondition } from '../PreflightTest';
import { PreflightTestReport } from '../../../../types';

export function getSingleNetworkCondition(
  stat: number | undefined,
  yellowThreshold: number,
  redThreshold: number,
) {
  if (typeof stat === 'undefined') {
    // We ignore values that are missing
    return NetworkCondition.Green;
  }

  if (stat >= redThreshold) return NetworkCondition.Red;
  if (stat >= yellowThreshold) return NetworkCondition.Yellow;
  return NetworkCondition.Green;
}

export default function getNetworkCondition(testReport?: PreflightTestReport) {
  if (!testReport) return undefined;

  const latency = testReport.stats.rtt?.average;
  const jitter = testReport.stats.jitter.average;
  const packetLoss = testReport.stats.packetLoss.average;

  return Math.min(
    getSingleNetworkCondition(latency, 200, 400),
    getSingleNetworkCondition(jitter, 30, 100),
    getSingleNetworkCondition(packetLoss, 3, 7),
  ) as NetworkCondition;
}
