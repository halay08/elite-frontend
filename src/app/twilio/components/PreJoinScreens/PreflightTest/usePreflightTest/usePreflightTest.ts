import { PreflightTestReport, PreflightTest } from '../../../../types';
import { TEST_DURATION } from '../PreflightTest';
import Video from 'twilio-video';
import { useEffect, useState, useRef } from 'react';
import useVideoContext from '../../../../hooks/useVideoContext/useVideoContext';

export default function usePreflightTest(
  publisherToken?: string,
  subscriberToken?: string,
) {
  const { isConnecting } = useVideoContext();
  const [testReport, setTestReport] = useState<PreflightTestReport>();
  const [testFailure, setTestFailure] = useState<Error>();
  const [isTestRunning, setIsTestRunning] = useState(false);
  const preflightTestRef = useRef<PreflightTest>();

  // This will stop the preflight test when the user connects to a room
  useEffect(() => {
    if (isConnecting) {
      preflightTestRef.current?.stop();
      preflightTestRef.current?.removeAllListeners('completed');
      preflightTestRef.current?.removeAllListeners('failed');
    }
  }, [isConnecting]);

  // This will stop the preflight test when the component is unmounted.
  useEffect(() => {
    return () => {
      preflightTestRef.current?.stop();
      preflightTestRef.current?.removeAllListeners('completed');
      preflightTestRef.current?.removeAllListeners('failed');
    };
  }, []);

  useEffect(() => {
    if (
      publisherToken &&
      subscriberToken &&
      !testReport &&
      !testFailure &&
      !isTestRunning
    ) {
      setIsTestRunning(true);

      preflightTestRef.current = Video.testPreflight(
        publisherToken,
        subscriberToken,
        { duration: TEST_DURATION },
      );

      preflightTestRef
        .current!.on('completed', (report: any) => {
          setTestReport(report);
          setIsTestRunning(false);
          console.log(
            'Preflight network test completed. See test report below:',
          );
          console.log(report);
        })
        .on('failed', (error: any) => {
          setTestFailure(error);
          setIsTestRunning(false);
          console.log('Preflight network test failed. See error below:');
          console.log(error);
        });
    }
  }, [publisherToken, subscriberToken, testReport, testFailure, isTestRunning]);

  return {
    testReport,
    testFailure,
    isTestRunning,
  };
}
