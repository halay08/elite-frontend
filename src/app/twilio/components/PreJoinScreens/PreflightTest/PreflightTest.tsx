import React from 'react';
import usePreflightTest from './usePreflightTest/usePreflightTest';
import { makeStyles, Typography, Grid } from '@material-ui/core';
import useGetPreflightTokens from './useGetPreflightTokens/useGetPreflightTokens';
import getNetworkCondition from './getNetworkCondition/getNetworkCondition';
import ProgressIndicator from './ProgressIndicator/ProgressIndicator';
import { SuccessIcon } from '../../../icons/SuccessIcon';
import WarningIcon from '../../../icons/WarningIcon';
import ErrorIcon from '../../../icons/ErrorIcon';

export const TEST_DURATION = 10000;

export enum NetworkCondition {
  Red,
  Yellow,
  Green,
}

const useStyles = makeStyles({
  iconContainer: {
    display: 'flex',
    alignItems: 'center',
    marginRight: '0.4em',
  },
  result: {
    color: 'white',
  },
});

function ResultItem({
  children,
  icon,
}: {
  children: React.ReactNode;
  icon: React.ReactNode;
}) {
  const classes = useStyles();

  return (
    <>
      <div className={classes.iconContainer}>{icon}</div>
      <Typography variant="subtitle2" className={classes.result}>
        {children}
      </Typography>
    </>
  );
}

export function Result({
  networkCondition,
  error,
}: {
  networkCondition?: NetworkCondition;
  error?: Error;
}) {
  if (error) {
    return (
      <ResultItem icon={<ErrorIcon />}>
        There was a problem connecting to the network
      </ResultItem>
    );
  }

  if (networkCondition === NetworkCondition.Red) {
    return (
      <ResultItem icon={<ErrorIcon />}>
        Poor network conditions. You may experience poor call quality and
        reliability.
      </ResultItem>
    );
  }

  if (networkCondition === NetworkCondition.Yellow) {
    return (
      <ResultItem icon={<WarningIcon />}>
        Poor network conditions. You may experience degraded video performance.
      </ResultItem>
    );
  }

  if (networkCondition === NetworkCondition.Green) {
    return (
      <ResultItem icon={<SuccessIcon />}>
        Your network connection is stable
      </ResultItem>
    );
  }

  return null;
}

export default function PreflightTest() {
  const classes = useStyles();

  const { tokens, tokenError } = useGetPreflightTokens();
  const { testFailure, testReport } = usePreflightTest(
    tokens?.[0],
    tokens?.[1],
  );

  const networkCondition = getNetworkCondition(testReport);

  return (
    <Grid
      container
      justify="center"
      alignItems="center"
      style={{ height: '55px', paddingBottom: '1.5em' }}
    >
      {!testFailure && !testReport ? (
        <>
          <ProgressIndicator />
          <Typography variant="subtitle2" className={classes.result}>
            Checking your network connection
          </Typography>
        </>
      ) : (
        <Result
          networkCondition={networkCondition}
          error={tokenError || testFailure}
        />
      )}
    </Grid>
  );
}
