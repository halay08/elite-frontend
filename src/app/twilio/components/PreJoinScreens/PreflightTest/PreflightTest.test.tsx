import React from 'react';
import getNetworkCondition from './getNetworkCondition/getNetworkCondition';
import PreflightTest, { NetworkCondition, Result } from './PreflightTest';
import ProgressIndicator from './ProgressIndicator/ProgressIndicator';
import { shallow } from 'enzyme';
import useGetPreflightTokens from './useGetPreflightTokens/useGetPreflightTokens';
import usePreflightTest from './usePreflightTest/usePreflightTest';

jest.mock('./useGetPreflightTokens/useGetPreflightTokens');
jest.mock('./usePreflightTest/usePreflightTest');
jest.mock('./getNetworkCondition/getNetworkCondition');

const mockUseGetPreflightTokens = useGetPreflightTokens as jest.Mock<any>;
const mockUsePreflightTest = usePreflightTest as jest.Mock<any>;
const mockGetNetworkCondition = getNetworkCondition as jest.Mock<any>;

mockUseGetPreflightTokens.mockImplementation(() => ({
  tokens: ['mockTokenA', 'mockTokenB'],
  tokenError: undefined,
}));
mockGetNetworkCondition.mockImplementation(() => NetworkCondition.Green);

describe('the PreflightTest component', () => {
  it('should render correctly when the test is in progress', () => {
    mockUsePreflightTest.mockImplementation(() => ({}));
    const wrapper = shallow(<PreflightTest />);
    expect(wrapper.text()).toContain('Checking your network connection');
    expect(wrapper.find(ProgressIndicator).exists()).toBe(true);
    expect(wrapper.find(Result).exists()).toBe(false);
  });

  it('should render correctly when there is a report', () => {
    mockUsePreflightTest.mockImplementation(() => ({
      testReport: 'mockTestReport',
    }));
    const wrapper = shallow(<PreflightTest />);
    expect(wrapper.find(ProgressIndicator).exists()).toBe(false);
    expect(wrapper.find(Result).exists()).toBe(true);
  });

  it('should render correctly when there is a failure', () => {
    mockUsePreflightTest.mockImplementation(() => ({
      testFailure: 'mockTestFailure',
    }));
    const wrapper = shallow(<PreflightTest />);
    expect(wrapper.find(ProgressIndicator).exists()).toBe(false);
    expect(wrapper.find(Result).exists()).toBe(true);
  });
});

describe('the Result component', () => {
  it('should render correctly when the network condition is green', () => {
    const wrapper = shallow(
      <Result networkCondition={NetworkCondition.Green} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when the network condition is yellow', () => {
    const wrapper = shallow(
      <Result networkCondition={NetworkCondition.Yellow} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when the network condition is red', () => {
    const wrapper = shallow(<Result networkCondition={NetworkCondition.Red} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when there is an error', () => {
    const wrapper = shallow(<Result error={new Error()} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not render anything when there are no props', () => {
    const wrapper = shallow(<Result />);
    expect(wrapper.isEmptyRender()).toBe(true);
  });
});
