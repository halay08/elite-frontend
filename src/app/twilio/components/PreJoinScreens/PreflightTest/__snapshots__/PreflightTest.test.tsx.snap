// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Result component should render correctly when the network condition is green 1`] = `
<ResultItem
  icon={<SuccessIcon />}
>
  Your network connection is stable
</ResultItem>
`;

exports[`the Result component should render correctly when the network condition is red 1`] = `
<ResultItem
  icon={<ErrorIcon />}
>
  Poor network conditions. You may experience poor call quality and reliability.
</ResultItem>
`;

exports[`the Result component should render correctly when the network condition is yellow 1`] = `
<ResultItem
  icon={<WarningIcon />}
>
  Poor network conditions. You may experience degraded video performance.
</ResultItem>
`;

exports[`the Result component should render correctly when there is an error 1`] = `
<ResultItem
  icon={<ErrorIcon />}
>
  There was a problem connecting to the network
</ResultItem>
`;
